### ISSUE TITLE
Consumer.Seek or Consumer.Assign having issues

### ISSUE THREAD

**Issue by Tang8330** [confluentinc/confluent-kafka-go#888]
Description
===========
This issue seems to have cropped up a couple of times, see:
* https://github.com/confluentinc/confluent-kafka-go/issues/121
* https://github.com/confluentinc/confluent-kafka-go/issues/748

In summary, if I try to do `consumer.Seek(...)` or `consumer.Assign(...)` to a `TopicPartition` where the offset doesn't exist anymore (maybe because the offset was older than retention); instead of having `Seek` or `Assign` error out, it returns `nil` and acts as if there was not an error. Meanwhile, I see a log line in my terminal.

```
%4|1667447031.006|OFFSET|rdkafka#consumer-1| [thrd:main]: topic_name [0]: offset reset (at offset 17011, broker 2) to END: fetch failed due to requested offset not available on the broker: Broker: Offset out of range
```

How to reproduce
================
* Try to run Seek or Assign on an offset that doesn't exist in the Broker anymore

Checklist
=========
Please provide the following information:

 - [x] confluent-kafka-go and librdkafka version (`LibraryVersion()`): v1.9.2
 - [x] Apache Kafka broker version: (Not sure how to check, I'm using Confluent Cloud)
 - [x] Client configuration: `ConfigMap{...}`
```go
	consumer, err := kafka.NewConsumer(&kafka.ConfigMap{
		"bootstrap.servers": config.GetSettings().Config.Kafka.BootstrapServer,
		"sasl.mechanisms":   "PLAIN",
		"security.protocol": "SASL_SSL",
		"group.id":          config.GetSettings().Config.Kafka.GroupID,
		"sasl.username":     config.GetSettings().Config.Kafka.Username,
		"sasl.password":     config.GetSettings().Config.Kafka.Password,
	})
```

 - [x] Operating system: Macbook M2 Monterey
 - [ ] Provide client logs (with `"debug": ".."` as necessary)
 - [ ] Provide broker log excerpts
 - [x] Critical issue

**Comment by edenhill** [comment]
Seek() and Assign() are asynchronous and simply tell the consumer where to start fetching, the calls will not wait for the FetchRequests to be made to the broker.
If you set `"auto.offset.reset": "error"` you'll receive a consumer error later when the fetch has been attempted.

**Comment by Tang8330** [comment]
@edenhill That's a bit confusing. I thought it was async when `timeoutMs` was 0? See: https://github.com/confluentinc/confluent-kafka-go/blob/cba5eb13d710dee51f4ecf95ea399dece014a513/kafka/consumer.go#L309-L322

Is there not a way to store the offset outside of Kafka and upon consumer restart to run Seek or Assign? Or is the only way to leverage the ConfigMap? (b/c I can just disable auto commit offset to be explicit).

**Comment by edenhill** [comment]
That docstring is incorrect, unfortunately. It'll only wait for the desire to seek to be registered, which is pretty much immediate anyway, so the timeout doesn't really do much.

**Comment by Tang8330** [comment]
Got it. @edenhill would you be able to confirm the workaround?

I essentially set `enable.auto.commit` to be false and only commit when my process is truly done.

TL;DR I have a main consumer process that takes messages of Kafka, does some mutation & put this in memory and then flush every X seconds or when Y messages build up. I'm trying to avoid message loss when the process creates, so I'm now committing after a successful flush.

**Comment by edenhill** [comment]
If your input and output are Kafka I'd recommend using transactions to get exactly once semantics

**Comment by Tang8330** [comment]
Oo interesting. Never used Kafka transactions before. I'll take a look and see how that might fit in with my workload.

Thank you 🥇

**Comment by mhowlett** [comment]
triaging as bug/docs - we need to update the docstring.

**Comment by milindl** [comment]
Closing this since the docstring is now updated.
